tests/evm/test_returndatacopy.py

```diff
@@ -0,0 +1,158 @@
+    CallContextFieldTag,
+    ExecutionState,
+    Opcode,
+    RW,
+    RWDictionary,
+    RWTableTag,
+    StepState,
+    Tables,
+    verify_steps,
+    CopyCircuit,
+    CopyDataTypeTag,
+)
+from zkevm_specs.copy_circuit import verify_copy_table
+from zkevm_specs.util import (
+    rand_fq,
+    rand_bytes,
+    memory_word_size,
+    memory_expansion,
+    GAS_COST_COPY,
+    MAX_N_BYTES_COPY_TO_MEMORY,
+    MEMORY_EXPANSION_QUAD_DENOMINATOR,
+    MEMORY_EXPANSION_LINEAR_COEFF,
+    RLC,
+)
+
+CALL_ID = 1
+CALLEE_MEMORY = [0x00] * 32 + [0x11] * 32
+TESTING_DATA = (
+    # simple cases
+    (0, 0, 32, 0, 32),
+    (100, 0, 32, 0, 32),
+    (0, 0, 32, 100, 32),
+    (100, 31, 1, 100, 32),
+)
+
+
+@pytest.mark.parametrize(
+    "dest_offset, offset, size, return_data_offset, return_data_length", TESTING_DATA
+)
+def test_returndatacopy(
+    dest_offset: int,
+    offset: int,
+    size: int,
+    return_data_offset: int,
+    return_data_length: int,
+):
+    randomness = rand_fq()
+
+    dest_offset_rlc = RLC(dest_offset, randomness)
+    memory_offset_rlc = RLC(offset, randomness)
+    size_rlc = RLC(size, randomness)
+
+    code = (
+        Bytecode()
+        .push32(size_rlc)
+        .push32(memory_offset_rlc)
+        .push32(dest_offset_rlc)
+        .returndatacopy()
+        .stop()
+    )
+    code_hash = RLC(code.hash(), randomness)
+
+    # assume return data is at the current largest memory pos
+    curr_mem_size = memory_word_size(return_data_offset + return_data_length)
+    next_mem_size, memory_gas_cost = memory_expansion(curr_mem_size, dest_offset + size)
+    gas = (
+        Opcode.RETURNDATACOPY.constant_gas_cost()
+        + memory_gas_cost
+        + memory_word_size(size) * GAS_COST_COPY
+    )
+    rw_dictionary = (
+        RWDictionary(1)
+        .stack_read(CALL_ID, 1021, dest_offset_rlc)
+        .stack_read(CALL_ID, 1022, memory_offset_rlc)
+        .stack_read(CALL_ID, 1023, size_rlc)
+        .call_context_read(
+            CALL_ID, CallContextFieldTag.LastCalleeReturnDataLength, return_data_length
+        )
+        .call_context_read(
+            CALL_ID, CallContextFieldTag.LastCalleeReturnDataOffset, return_data_offset
+        )
+    )
+    # rw counter before memory writes
+    rw_counter_interim = rw_dictionary.rw_counter
+    steps = [
+        StepState(
+            execution_state=ExecutionState.RETURNDATACOPY,
+            rw_counter=1,
+            call_id=CALL_ID,
+            is_root=True,
+            code_hash=code_hash,
+            program_counter=99,
+            stack_pointer=1021,
+            memory_size=curr_mem_size,
+            gas_left=gas,
+        ),
+    ]
+
+    src_data = dict(
+        [
+            (i, CALLEE_MEMORY[i] if i < len(CALLEE_MEMORY) else 0)
+            for i in range(return_data_offset, return_data_offset + return_data_length)
+        ]
+    )
+    copy_circuit = CopyCircuit().copy(
+        randomness,
+        rw_dictionary,
+        CALL_ID,
```

Comment:
Need to create a mock `CALLEE_ID` for the test.

```suggestion
        CALLEE_ID,
```